### PR TITLE
Fix pv's objectsync spec.ObjectKind is empty

### PR DESCRIPTION
**Pull request by neiba**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 1]
```diff
@@ -1004,6 +1008,12 @@ func NewUpstreamController(config *v1alpha1.EdgeController, factory k8sinformer.
 	uc.podLister = factory.Core().V1().Pods().Lister()
 	uc.configMapLister = factory.Core().V1().ConfigMaps().Lister()
 	uc.secretLister = factory.Core().V1().Secrets().Lister()
+	factory.Core().V1().PersistentVolumes()
+	uc.pvLister = factory.Core().V1().PersistentVolumes().Lister()
+	factory.Core().V1().PersistentVolumeClaims()
+	uc.pvcLister = factory.Core().V1().PersistentVolumeClaims().Lister()
+	factory.Storage().V1().VolumeAttachments()
+	uc.attachLister = factory.Storage().V1().VolumeAttachments().Lister()
```
Comment (Iceber) [inline cloud/pkg/edgecontroller/controller/upstream.go:1016]: It seems that you just need to call `factory...Lister()` directly.

File: cloud/pkg/cloudhub/handler/messagehandler.go [hunk 2]
```diff
@@ -561,6 +561,10 @@ func (mh *MessageHandle) saveSuccessPoint(msg *beehiveModel.Message, info *model
 					ObjectName:       resourceName,
 				},
 			}
+			if objectSync.Spec.ObjectKind == "" {
+				klog.Errorf("Failed to init objectSync: %s, ObjectKind is empty, msg content: %v", objectSyncName, msg.GetContent())
```
Comment (fisherxu) [inline cloud/pkg/cloudhub/handler/messagehandler.go:565]: I'm wondering which message's kind/type will be empty?

File: cloud/pkg/edgecontroller/controller/upstream.go (hunk 1 above)
Comment (fisherxu) [reply cloud/pkg/edgecontroller/controller/upstream.go:1016]: +1

File: cloud/pkg/synccontroller/objectsync.go [hunk 3]
```diff
@@ -90,8 +86,9 @@ func sendEvents(err error, nodeName string, sync *v1alpha1.ObjectSync, resourceT
 	if CompareResourceVersion(objectResourceVersion, sync.Status.ObjectResourceVersion) > 0 {
 		// trigger the update event
 		klog.V(4).Infof("The resourceVersion: %s of %s in K8s is greater than in edgenode: %s, send the update event", objectResourceVersion, resourceType, sync.Status.ObjectResourceVersion)
-		msg := buildEdgeControllerMessage(nodeName, sync.Namespace, resourceType, sync.Spec.ObjectName, model.UpdateOperation, obj)
-		beehiveContext.Send(commonconst.DefaultContextSendModuleName, *msg)
+		if msg := buildEdgeControllerMessage(nodeName, sync.Namespace, resourceType, sync.Spec.ObjectName, model.UpdateOperation, obj); msg != nil {
+			beehiveContext.Send(commonconst.DefaultContextSendModuleName, *msg)
+		}
```
Comment (fisherxu) [inline cloud/pkg/synccontroller/objectsync.go:91]: Does the modification in this file related to the issue this pr fixed? If not, we can move it to another PR :)

File: cloud/pkg/edgecontroller/controller/upstream.go (hunk 1 above)
Comment (fisherxu) [reply cloud/pkg/edgecontroller/controller/upstream.go:1016]: +1

File: cloud/pkg/cloudhub/handler/messagehandler.go (hunk 2 above)
Comment (neiba) [reply cloud/pkg/cloudhub/handler/messagehandler.go:565]: i made a mistake, configmap object metatype is set is edgecontroller downstream: https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/edgecontroller/manager/common.go#L29

File: cloud/pkg/synccontroller/objectsync.go (hunk 3 above)
Comment (neiba) [reply cloud/pkg/synccontroller/objectsync.go]: i found related PR had been merged: https://github.com/kubeedge/kubeedge/pull/3570

File: cloud/pkg/cloudhub/handler/messagehandler.go (hunk 2 above)
Comment (fisherxu) [reply cloud/pkg/cloudhub/handler/messagehandler.go:565]: I'm wondering what kind of message's ObjectKind will be empty?

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 4]
```diff
@@ -572,6 +574,9 @@ func kubeClientGet(uc *UpstreamController, namespace string, name string, queryT
 		klog.Error(err)
 		return nil, err
 	}
+	if err == nil && obj != nil {
+		util.SetMetaType(obj.(runtime.Object))
```
Comment (fisherxu) [inline cloud/pkg/edgecontroller/controller/upstream.go:578]: Where will use the Meta(apiversion and kind)?

File: cloud/pkg/synccontroller/objectsync.go [hunk 5]
```diff
@@ -44,8 +44,13 @@ func (sctl *SyncController) manageObject(sync *v1alpha1.ObjectSync) {
 		newObject.SetNamespace(sync.Namespace)
 		newObject.SetName(sync.Spec.ObjectName)
 		newObject.SetUID(types.UID(getObjectUID(sync.Name)))
-		msg := buildEdgeControllerMessage(nodeName, sync.Namespace, resourceType, sync.Spec.ObjectName, model.DeleteOperation, newObject)
-		beehiveContext.Send(commonconst.DefaultContextSendModuleName, *msg)
+		if msg := buildEdgeControllerMessage(nodeName, sync.Namespace, resourceType, sync.Spec.ObjectName, model.DeleteOperation, newObject); msg != nil {
```
Comment (fisherxu) [inline cloud/pkg/synccontroller/objectsync.go:47]: Will the msg be nil? Which kind of message?

**Review by fisherxu — COMMENTED** [review]
Just some questions :)

File: cloud/pkg/edgecontroller/controller/upstream.go (hunk 4 above)
Comment (neiba) [reply cloud/pkg/edgecontroller/controller/upstream.go:578]: cloudhub send msg to edge successful and then create objectsync, which will use msg resource meta
https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/cloudhub/handler/messagehandler.go#L560

File: cloud/pkg/synccontroller/objectsync.go (hunk 5 above)
Comment (neiba) [reply cloud/pkg/synccontroller/objectsync.go:47]: if objectsync's spec.ObjectKind is empty, then buildEdgeControllerMessage would return nil, such as pvc's objectsync

Comment (fisherxu) [reply cloud/pkg/synccontroller/objectsync.go:47]: Can you explain why the ObjectKind of pvc's objectsync is empty?

File: cloud/pkg/edgecontroller/controller/upstream.go (hunk 4 above)
Comment (fisherxu) [reply cloud/pkg/edgecontroller/controller/upstream.go:578]: And can we get the obj from lister in L563 565 567?

File: cloud/pkg/synccontroller/objectsync.go (hunk 5 above)
Comment (neiba) [reply cloud/pkg/synccontroller/objectsync.go:47]: 1. edged query a pvc
2. edgecontroller upstream receive the msg
3. edgecontroller upstream get pvc from k8s api 
4. edgecontroller store pvc object to msg.content, and send msg to cloudhub
5. cloudhub send msg to edgehub
6. cloudhub get response from edge, and create pvc's objectsync
7. objectsync.spec.objectkind = util.GetMessageResourceType(msg)

step 7 msg and step 4 msg are same one, and msg.content is pvc object, which metatype is empty. So we need SetMetaType

Comment (fisherxu) [reply cloud/pkg/synccontroller/objectsync.go:47]: OK, get it, thanks! And if the msg is empty here, is it reasonable to delete the objectsync as below?

Comment (fisherxu) [reply cloud/pkg/synccontroller/objectsync.go:47]: And if we add the `util.SetMetaType(obj.(runtime.Object))`, maybe the msg will not be empty? :)

Comment (neiba) [reply cloud/pkg/synccontroller/objectsync.go:47]: there are three cases:
1. objectsync's resource is deleted
2. objectsync's resource is deleted, but a same name resource created
3. objectsync.Spec.ObjectKind is empty.

case 1: program will entry line 48, send delete msg to edge and then delete objectsync
case 2: program will entry line 73, sendEvents function will send delete msg to edge and then delete objectsync
case 3: program will entry line 50, delete objectsync directly

maybe msg will not be empty, add a check just for comfort. ：）

File: cloud/pkg/synccontroller/objectsync.go [hunk 6]
```diff
@@ -77,8 +82,9 @@ func sendEvents(err error, nodeName string, sync *v1alpha1.ObjectSync, resourceT
 	if err != nil && apierrors.IsNotFound(err) {
 		//trigger the delete event
 		klog.Infof("%s: %s has been deleted in K8s, send the delete event to edge", resourceType, sync.Spec.ObjectName)
-		msg := buildEdgeControllerMessage(nodeName, sync.Namespace, resourceType, sync.Spec.ObjectName, model.DeleteOperation, obj)
-		beehiveContext.Send(commonconst.DefaultContextSendModuleName, *msg)
+		if msg := buildEdgeControllerMessage(nodeName, sync.Namespace, resourceType, sync.Spec.ObjectName, model.DeleteOperation, obj); msg != nil {
+			beehiveContext.Send(commonconst.DefaultContextSendModuleName, *msg)
```
Comment (neiba) [inline cloud/pkg/synccontroller/objectsync.go:86]: not delete objectsync when msg == nil, because:
1. not crdclient :)
2. msg won't be nil almostly, add check msg == nil to avoid cloudcore crash when beehive send a nil msg.

File: cloud/pkg/synccontroller/objectsync.go (hunk 5 above)
Comment (fisherxu) [reply cloud/pkg/synccontroller/objectsync.go:47]: Make sense, thanks for the clarification!

File: cloud/pkg/cloudhub/handler/messagehandler.go [hunk 7]
```diff
@@ -561,6 +561,10 @@ func (mh *MessageHandle) saveSuccessPoint(msg *beehiveModel.Message, info *model
...
+				return
```
Comment (fisherxu) [inline cloud/pkg/cloudhub/handler/messagehandler.go:566]: Last question maybe, with `util.SetMetaType(obj.(runtime.Object))` in kubeClientGet, I think the kind will almost never empty, right?

**Review by fisherxu — COMMENTED** [review]
@My-Pleasure Would you like to take a look?

File: cloud/pkg/cloudhub/handler/messagehandler.go (hunk 7 above)
Comment (neiba) [reply cloud/pkg/cloudhub/handler/messagehandler.go:566]: yep, i think so. Dynamiccontroller informers alse already called SetMetaType in eventhandler. Cloudhub send other type msgs to edge by calling ListMessageWriteLoop, which no need to create objectsync, such as device twin, applicationResponse, user data.

Comment (fisherxu) [reply cloud/pkg/cloudhub/handler/messagehandler.go:566]: Sounds good.

**Comment by My-pleasure** [comment]
I agree with adding `util.SetMetaType` in function `kubeClientGet`. It's good.
By the way, when upstream controller gets the obj from lister, that obj doesn't need to `SetMetaType`, like ConfigMap. Only pv and pvc .etc which get by kubeClient need to `SetMetaType`. And if we want get pv from lister, we should set a pv lister in upstream controller which is not had now.

**Comment by fisherxu** [comment]
> By the way, when upstream controller gets the obj from lister, that obj doesn't need to SetMetaType, like ConfigMap. Only pv and pvc .etc which get by kubeClient need to SetMetaType. And if we want get pv from lister, we should set a pv lister in upstream controller which is not had now.

Just have a check, the obj in list also has no MetaType info, so we need to set the MetaType. @My-pleasure

**Comment by neiba** [comment]
> I agree with adding `util.SetMetaType` in function `kubeClientGet`. It's good. By the way, when upstream controller gets the obj from lister, that obj doesn't need to `SetMetaType`, like ConfigMap. Only pv and pvc .etc which get by kubeClient need to `SetMetaType`. And if we want get pv from lister, we should set a pv lister in upstream controller which is not had now.

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Comment by fisherxu** [comment]
@neiba Could you please cherry-pick this pr to release-1.9 branch using script `hack/cherry_pick_pull.sh`? :)
